### PR TITLE
Improve health check to return more granular details

### PR DESCRIPTION
**Pull request by sd2k**
The Dashboard AI team have requested that the plugin health check returns more details about whether the OpenAI functionality is configured, and whether that configuration is valid. This PR adds it, and similar functionality for the vector service.

It's backwards incompatible with past health checks unfortunately, but the llmclient package does include a change which should be compatible with both the current and previous version of the plugin.

We're definitely due a refactor of the way we handle OpenAI config and requests, but I think that will be part of #80.

### CODE REVIEW

**Comment by sd2k** [comment]
TODO:

- cache the health check result on the `App` struct, since it's expensive and slow and the app will be recreated if the plugin settings change anyway.
- Add a 'test' button to the settings UI which runs a health check (bypassing the cache?)

File: src/components/AppConfig/HealthCheck.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+  severity = getAlertSeverity(props.status ?? 'error', props.details);
+  const message = severity === 'success' ? 'Health check succeeded!' : props.message;
+
+  return (
+    <div className="gf-form-group p-t-2">
+      <Alert severity={severity} title={message}>
+        <ShowOpenAIHealth openAI={props.details.openAI} />
+        <ShowVectorHealth vector={props.details.vector} />
+      </Alert>
```
Comment (ivanortegaalba) [inline src/components/AppConfig/HealthCheck.tsx:76]: Thinking about the scenario where the OpenAI is correctly configured but vectorDB is ignored or misconfigured, probably it would be clearer to have two alerts, one per technology. WDYT?

Comment (ivanortegaalba) [inline src/components/AppConfig/HealthCheck.tsx:76]: form-group can probably be replaced by a VerticalGroup. I think we're trying to deprecate those classes in favor of components

File: src/components/AppConfig/HealthCheck.tsx [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  openAI: OpenAIHealthDetails | boolean;
+  vector: VectorHealthDetails | boolean;
+  version: string;
+}
+
+interface OpenAIHealthDetails {
+  configured: boolean;
+  ok: boolean;
+  error?: string;
+  models: Record<string, OpenAIModelHealthDetails>;
```
Comment (ivanortegaalba) [inline src/components/AppConfig/HealthCheck.tsx:16]: It could be great to have some comments here explaining the difference between ok and configured. WDYT?

File: src/components/AppConfig/HealthCheck.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  }
+
+  severity = getAlertSeverity(props.status ?? 'error', props.details);
+  const message = severity === 'success' ? 'Health check succeeded!' : props.message;
+
+  return (
+    <div className="gf-form-group p-t-2">
+      <Alert severity={severity} title={message}>
+        <ShowOpenAIHealth openAI={props.details.openAI} />
+        <ShowVectorHealth vector={props.details.vector} />
```
Comment (ivanortegaalba) [inline src/components/AppConfig/HealthCheck.tsx:75]: Do we need to show both checks if one of them is not configured? Maybe, if it is not configured, we can just hide it, to not mislead the user.

File: pkg/plugin/health_test.go [hunk 4]
```diff
@@ -87,10 +137,32 @@ func TestCheckHealth(t *testing.T) {
...
+				},
+			},
+			expDetails: healthCheckDetails{
+				OpenAI: openAIHealthDetails{
+					Configured: true,
+					OK:         true,
+					Error:      "",
+					Models: map[string]openAIModelHealth{
+						"gpt-3.5-turbo": {OK: true, Error: ""},
+						"gpt-4":         {OK: false, Error: `unexpected status code: 404: {"error": "model does not exist"}`},
```
Comment (ivanortegaalba) [inline pkg/plugin/health_test.go:158]: If we return the errors like `foo: bar: {error: baz}` it is hard to parse in the frontend. Could we just return an object? is this any type of convention?
I'd love to return to the frontend as much information as possible from OpenAI since this is part of the configuration phase.

This is the best we can do when the request fails when using the query:

File: src/components/AppConfig/HealthCheck.tsx [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          {Object.entries(openAI.models).map(([model, details], i) => (
+            <tr key={i}>
+              <td>{model}: </td>
+              <td>{details.ok ? 'OK' : `Error: ${details.error}`}</td>
+            </tr>
+          ))}
+        </tbody>
+      </table>
+    </div>
+  )
```
Comment (ivanortegaalba) [inline src/components/AppConfig/HealthCheck.tsx:103]: Instead of using a table, what do you think about using just a bullet list with `li`? It looks like a lot of information on a corner.

Captura de pantalla 2023-10-04 a las 12 25 45

Also, I'd recommend having the banner above the button. You can miss it if you don't scroll.

File: src/components/AppConfig/AppConfig.tsx [hunk 6]
```diff
@@ -86,18 +92,24 @@ export const AppConfig = ({ plugin }: AppConfigProps) => {
...
+            const result = await checkPluginHealth(plugin.meta.id);
+            setIsUpdating(false);
+            setHealthCheck(result.data);
           }}
           disabled={!updated}
         >
-          Save settings
+          Save &amp; test
         </Button>
+        {isUpdating && <Spinner />}
```
Comment (ivanortegaalba) [inline src/components/AppConfig/AppConfig.tsx:109]: Instead of rendering a spinner out of the button, you can use [LoadingPlaceholder](https://developers.grafana.com/ui/latest/index.html?path=/story/general-loadingplaceholder--basic) to replace the button content while loading.

File: src/components/AppConfig/HealthCheck.tsx (hunk 1 above)
Comment (sd2k) [reply src/components/AppConfig/HealthCheck.tsx]: Thanks!

File: src/components/AppConfig/HealthCheck.tsx (hunk 2 above)
Comment (sd2k) [reply src/components/AppConfig/HealthCheck.tsx]: Good point, done.

File: src/components/AppConfig/HealthCheck.tsx (hunk 3 above)
Comment (sd2k) [reply src/components/AppConfig/HealthCheck.tsx]: Good point, done! 🙂

File: src/components/AppConfig/HealthCheck.tsx (hunk 1 above)
Comment (sd2k) [reply src/components/AppConfig/HealthCheck.tsx]: Yep, makes sense and looks nicer, thanks!

File: src/components/AppConfig/HealthCheck.tsx (hunk 5 above)
Comment (sd2k) [reply src/components/AppConfig/HealthCheck.tsx]: Agreed, I've added that now.

File: src/components/AppConfig/AppConfig.tsx (hunk 6 above)
Comment (sd2k) [reply src/components/AppConfig/AppConfig.tsx]: I've done something similar in the latest commit - take a look and see what you think?

Comment (sd2k) [reply src/components/AppConfig/AppConfig.tsx]: There might be existing UI patterns that I'm not aware of here but I haven't noticed buttons disappearing while things are loading 🤔

**Review by ivanortegaalba — APPROVED** [review]
✅ The experience configuring the plugin is great. The right place to have debug info.
✅ The feature still working as expected
✅ The health check works in every scenario
